### PR TITLE
don't remove resource output when the content type is resource and copyResources is false

### PR DESCRIPTION
**Pull request by 928PJY**
when there is some error happen in the resource file, and the `coprResource` config is false, we will delete the original resource file in the current build folder.

### CODE REVIEW

File: src/docfx/publish/PublishModelBuilder.cs [hunk 1]
```diff
@@ -66,7 +66,7 @@ public bool TryAdd(Document file, PublishItem item)
 
                     foreach (var conflictingFile in files.Keys)
                     {
-                        if (_publishItems.TryRemove(conflictingFile, out var item) && item.Path != null)
+                        if (_publishItems.TryRemove(conflictingFile, out var item) && item.Path != null && (conflictingFile.ContentType != ContentType.Resource || conflictingFile.Docset.Config.Output.CopyResources))
```
Comment (yufeih) [inline src/docfx/publish/PublishModelBuilder.cs]: check whether `item.Path` is inside output folder

**Comment by OsmondJiang** [comment]
what issue we are going to resolve here? would you please add it to description？

File: src/docfx/publish/PublishModelBuilder.cs [hunk 2]
```diff
@@ -93,7 +93,7 @@ public bool TryAdd(Document file, PublishItem item)
 
                 foreach (var conflictingFile in conflictingFiles)
                 {
-                    if (_publishItems.TryRemove(conflictingFile, out var item) && item.Path != null)
+                    if (_publishItems.TryRemove(conflictingFile, out var item) && item.Path != null && (conflictingFile.ContentType != ContentType.Resource || conflictingFile.Docset.Config.Output.CopyResources))
```
Comment (yufeih) [inline src/docfx/publish/PublishModelBuilder.cs]: @928PJY , can you change the test framework to assert that tests shouldn't change any files in source folder? You can simplify compare the last modified time of each file in the source repo.

File: test/docfx.Test/e2e/E2ETest.cs [hunk 3]
```diff
@@ -453,5 +458,18 @@ private static void VerifyLogsHasLineInfo(string[] logs)
...
+        private static void VerifyNoChangesOnInputFiles(FileInfo[] inputFiles, string[] skippableInputs)
+        {
+            foreach(var file in inputFiles)
+            {
+                if (skippableInputs.Contains(file.Name))
+                {
+                    continue;
+                }
+                Assert.True(File.Exists(file.FullName), $"Input file {file.Name} has been deleted");
+                Assert.True(file.LastWriteTime == new FileInfo(file.FullName).LastWriteTime, $"Input file {file.Name} has been changed");
```
Comment (live1206) [inline test/docfx.Test/e2e/E2ETest.cs]: Seems like the `LastWriteTime` assertion is always true?

File: src/docfx/publish/PublishModelBuilder.cs [hunk 4]
```diff
@@ -126,5 +126,10 @@ public bool TryAdd(Document file, PublishItem item)
 
             return (model, fileManifests);
         }
+
+        private bool IsDeletableFile(PublishItem item)
+        {
+            return !item.Path.StartsWith("..");
```
Comment (live1206) [inline src/docfx/publish/PublishModelBuilder.cs]: Is `item.Path` always a relative path?

**Comment by live1206** [comment]
add a test case

File: test/docfx.Test/e2e/E2ETest.cs (hunk 3 above)
Comment (928PJY) [reply test/docfx.Test/e2e/E2ETest.cs]: No, the `FileInfo` will not change once it has been created

---
In reply to: [284090364](https://github.com/dotnet/docfx/pull/4518#discussion_r284090364) [](ancestors = 284090364)

File: src/docfx/publish/PublishModelBuilder.cs (hunk 4 above)
Comment (928PJY) [reply src/docfx/publish/PublishModelBuilder.cs]: yes, related to manifest file

---
In reply to: [284090685](https://github.com/dotnet/docfx/pull/4518#discussion_r284090685) [](ancestors = 284090685)

Comment (yufeih) [reply src/docfx/publish/PublishModelBuilder.cs]: If the output path is in another windows directory, `item.Path` will look like `D:/a/b/c`, thus this check will fail. There is a `PathUtility.Match` method that tries to do the same thing, use that method and update the implementation?

File: test/docfx.Test/e2e/E2ETest.cs (hunk 3 above)
Comment (yufeih) [reply test/docfx.Test/e2e/E2ETest.cs]: @928PJY , better not relying on this implementation details, suggest using a `Dictionary<string, DateTime>`

File: test/docfx.Test/e2e/E2ETest.cs [hunk 5]
```diff
@@ -453,5 +458,18 @@ private static void VerifyLogsHasLineInfo(string[] logs)
...
+
+        private static void VerifyNoChangesOnInputFiles(FileInfo[] inputFiles, string[] skippableInputs)
+        {
+            foreach(var file in inputFiles)
+            {
+                if (skippableInputs.Contains(file.Name))
+                {
+                    continue;
+                }
+                Assert.True(File.Exists(file.FullName), $"Input file {file.Name} has been deleted");
```
Comment (yufeih) [inline test/docfx.Test/e2e/E2ETest.cs]: Also need to assert that we don't generate new files in inputs, so

```csharp
Assert.True(inputs.OrderBy(path).SequentialEquals(outputs.OrderBy(path)))
```

File: src/docfx/publish/PublishModelBuilder.cs [hunk 6]
```diff
@@ -126,5 +126,10 @@ public bool TryAdd(Document file, PublishItem item)
 
             return (model, fileManifests);
         }
+
+        private bool IsDeletableFile(PublishItem item)
```
Comment (yufeih) [inline src/docfx/publish/PublishModelBuilder.cs]:
```suggestion
        private bool IsInsideOutputPath(PublishItem item)
```

File: test/docfx.Test/e2e/E2ETest.cs (hunk 3 above)
Comment (928PJY) [reply test/docfx.Test/e2e/E2ETest.cs]: sure

---
In reply to: [284527224](https://github.com/dotnet/docfx/pull/4518#discussion_r284527224) [](ancestors = 284527224)

File: test/docfx.Test/e2e/E2ETest.cs [hunk 7]
```diff
@@ -453,5 +458,26 @@ private static void VerifyLogsHasLineInfo(string[] logs)
...
+                                .Select(file => file.FullName)
+                                .Except(dir.GetFiles($"_site/*", SearchOption.AllDirectories).Select(file => file.FullName)).ToList();
+            foreach (var (file, lastModifyTime) in inputFiles)
+            {
+                var currentFile = new FileInfo(file);
+                if (skippableInputs.Contains(currentFile.Name))
+                {
+                    continue;
+                }
+                Assert.True(File.Exists(file), $"Input file {currentFile.Name} has been deleted");
```
Comment (yufeih) [inline test/docfx.Test/e2e/E2ETest.cs]: file.Exists
```suggestion
                Assert.True(currentFile.Exists, $"Input file {currentFile.Name} has been deleted");
```

**Pull request merged**